Signal/src/ViewControllers/MessageMetadataViewController.swift

```diff
@@ -0,0 +1,393 @@
+            return rows
+        }
+        self.attachmentStream = attachmentStream
+
+        if let filePath = attachmentStream.filePath() {
+            dataSource = DataSourcePath.dataSource(withFilePath:filePath)
+        }
+
+        guard let dataSource = dataSource else {
+            rows.append(valueRow(name: NSLocalizedString("MESSAGE_METADATA_VIEW_ATTACHMENT_MISSING_FILE",
+                                                         comment: "Label for 'missing' attachments in the 'message metadata' view."),
+                                 value:""))
+            return rows
+        }
+
+        let fileSize = dataSource.dataLength()
+        rows.append(valueRow(name: NSLocalizedString("MESSAGE_METADATA_VIEW_ATTACHMENT_FILE_SIZE",
+                                                     comment: "Label for file size of attachments in the 'message metadata' view."),
+                             value:ViewControllerUtils.formatFileSize(UInt(fileSize))))
+
+        if let dataUTI = MIMETypeUtil.utiType(forMIMEType:contentType) {
+            if attachment.isVoiceMessage() {
+                rows.append(valueRow(name: NSLocalizedString("MESSAGE_METADATA_VIEW_VOICE_MESSAGE",
+                                                             comment: "Label for voice messages of the 'message metadata' view."),
+                                     value:""))
+            } else {
+                rows.append(valueRow(name: NSLocalizedString("MESSAGE_METADATA_VIEW_MEDIA",
+                                                             comment: "Label for media messages of the 'message metadata' view."),
+                                     value:""))
+            }
+            let attachment = SignalAttachment(dataSource : dataSource, dataUTI: dataUTI)
+            let mediaMessageView = MediaMessageView(attachment:attachment)
+            self.mediaMessageView = mediaMessageView
+            rows.append(mediaMessageView)
+        }
+        return rows
+    }
+
+    private func recipientStatus(forOutgoingMessage message: TSOutgoingMessage, recipientId: String) -> String {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateStyle = .short
+        dateFormatter.timeStyle = .long
+
+        let recipientReadMap = message.recipientReadMap
+        if let readTimestamp = recipientReadMap[recipientId] {
+            assert(message.messageState == .sentToService)
+            let readDate = NSDate.ows_date(withMillisecondsSince1970:readTimestamp.uint64Value)
+            return String(format:NSLocalizedString("MESSAGE_STATUS_READ_WITH_TIMESTAMP_FORMAT",
+                                                   comment: "message status for messages read by the recipient. Embeds: {{the date and time the message was read}}."),
+                          dateFormatter.string(from:readDate))
+        }
+
+        // TODO: We don't currently track delivery state on a per-recipient basis.
+        //       We should.
```

Comment:
This TODO is addressed here: https://github.com/WhisperSystems/Signal-iOS/pull/2564